File: spider.py
import sys
import requests
from bs4 import BeautifulSoup
from typing import List

URL = 'http://www.waptxt.org/96031'
WEBSITE = 'http://www.waptxt.org'
ENCODING = 'gbk'
PARSER = 'lxml'


class Spider:

    def __init__(self, website: str, url: str, encoding: str, parser: str) -> None:
        '''
        @param website 小说网站地址
        @param url 小说页面（目录页面）地址
        @param encoding: 网站编码
        @param parser: 解码器
        '''

        self.website = website
        self.url = url
        self.encoding = encoding
        self.parser = parser

    def get_bsobj(self, url: str) -> BeautifulSoup:
        response = requests.get(url)
        response.encoding = self.encoding
        return BeautifulSoup(response.text, self.parser)

    def get_chapter(self, curl: str) -> str:
        '''
        获取章节名称及其内容

        curl: 章节地址
        return: 章节名称\\n章节内容\\n\\n

        其中，章节内容是每段进行分行
        '''

        soup = self.get_bsobj(curl)
        cont = soup.find('div', {'class': 'con_txt'})
        title = soup.find('div', {'class': 'title'})

        title_text = title.h1.text.strip()
        cont_text = '\n'.join(cont.stripped_strings)

        return title_text + '\n' + cont_text + '\n\n'

    def get_chapter_url_list(self) -> List[str]:
        '''
        获取章节url的列表
        '''

        url = self.url
        res: list[str] = []

        while True:
            soup = self.get_bsobj(url)
            for i in soup.dl.findAll('a'):
                res.append(self.website + i.attrs['href'])

            next_page = soup.find('span', {'class': 'right'})
            next_page_href = next_page.a.attrs.get('href', None)
            if not next_page_href:
                break
            url = self.website + next_page_href

        return res

    def save_novel(self, file_path: str, is_append: bool, start_chapter: int = 1):
        '''
        保存小说

        @param file_path 保存文件路径
        @param is_append 是否追加
        @param start_chapter 从第几章开始
        '''

        print('开始获取章节目录...')
        chapter_url_list = self.get_chapter_url_list()
        print('章节目录获取完毕')
        print('开始缓存')
        with open(file_path, 'a' if is_append else 'w', encoding='utf-8') as f:
            for i, v in enumerate(chapter_url_list[start_chapter - 1:]):
                f.write(self.get_chapter(v))
                print(f'第{i + start_chapter}章缓存完毕')
        print('所有章节缓存完毕！')


if __name__ == '__main__':
    spider = Spider(WEBSITE, URL, ENCODING, PARSER)
    spider.save_novel(sys.path[0] + '/道诡异仙.txt', False)

File: rust/src/spider.rs
use anyhow::Result;
use scraper::{Html, Selector};

pub const DEFAULT_URL: &str = "http://www.waptxt.org/96031";
pub const DEFAULT_WEBSITE: &str = "http://www.waptxt.org";
pub const DEFAULT_ENCODING: &str = "gbk";
const USER_AGENT: &str = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";
const MAX_RETRIES: u32 = 100;

pub struct Spider {
    pub website: String,
    pub url: String,
    pub encoding: String,
    client: reqwest::Client,
}

impl Spider {
    pub fn new(website: String, url: String, encoding: String) -> Self {
        Spider {
            website,
            url,
            encoding,
            client: reqwest::Client::new(),
        }
    }

    pub async fn get_bsobj(&self, url: &str) -> Result<Html> {
        let response = self
            .client
            .get(url)
            .header("User-Agent", USER_AGENT)
            .send()
            .await?;

        let bytes = response.bytes().await?;

        let text = if self.encoding.eq_ignore_ascii_case("gbk") {
            let (cow, _, _) = encoding_rs::GBK.decode(&bytes);
            cow.into_owned()
        } else {
            String::from_utf8_lossy(&bytes).into_owned()
        };

        Ok(Html::parse_document(&text))
    }

    pub async fn get_chapter_url_list(&self) -> Result<Vec<String>> {
        let mut url = self.url.clone();
        let mut result = Vec::new();

        let dl_a_selector = Selector::parse("dl a").unwrap();
        let next_page_selector = Selector::parse("span.right a").unwrap();

        loop {
            let doc = self.get_bsobj(&url).await?;

            for element in doc.select(&dl_a_selector) {
                if let Some(href) = element.value().attr("href") {
                    result.push(format!("{}{}", self.website, href));
                }
            }

            let next_href = doc
                .select(&next_page_selector)
                .next()
                .and_then(|el| el.value().attr("href"))
                .map(|h| h.to_string());

            match next_href {
                Some(href) => url = format!("{}/{}", self.website, href),
                None => break,
            }
        }

        Ok(result)
    }

    pub async fn get_chapter(&self, curl: &str) -> Result<String> {
        let title_selector = Selector::parse("div.title h1").unwrap();
        let content_selector = Selector::parse("div.con_txt").unwrap();
        let next_selector = Selector::parse("div.chapter_go a#xiazhang").unwrap();

        let mut curl = curl.to_string();
        let mut title = String::new();
        let mut title_set = false;
        let mut texts: Vec<String> = Vec::new();

        loop {
            let doc = self.get_bsobj(&curl).await?;

            if !title_set {
                title = doc
                    .select(&title_selector)
                    .next()
                    .map(|el| el.text().collect::<String>().trim().to_string())
                    .ok_or_else(|| anyhow::anyhow!("未找到 div.title h1"))?;
                title_set = true;
            }

            let content = doc
                .select(&content_selector)
                .next()
                .ok_or_else(|| anyhow::anyhow!("未找到 div.con_txt"))?;
            let page_text = content
                .text()
                .map(|t| t.trim().to_string())
                .filter(|t| !t.is_empty())
                .collect::<Vec<_>>()
                .join("\n");
            texts.push(page_text);

            let next = doc
                .select(&next_selector)
                .next()
                .ok_or_else(|| anyhow::anyhow!("未找到 a#xiazhang"))?;

            let next_text: String = next.text().collect::<String>().trim().to_string();
            if next_text == "下一章" {
                break;
            }

            let href = next
                .value()
                .attr("href")
                .ok_or_else(|| anyhow::anyhow!("a#xiazhang 缺少 href"))?;
            let href = href.trim_start_matches('/');
            curl = format!("{}/{}", self.website, href);
        }

        Ok(format!("{}\n{}\n\n", title, texts.join("\n")))
    }

    pub async fn save_novel(
        &self,
        file_path: &str,
        is_append: bool,
        start_chapter: usize,
    ) -> Result<()> {
        todo!()
    }
}
